### PR TITLE
Milestone 3

### PR DESCRIPTION
**Pull request by RokoVarano**
Update:
Implemented changes to use method instead of enumerable/logic and defined associations with SQL query

### CODE REVIEW

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Hello team, you have shown a great effort so far, I applaud your energy 👏.

## To Improve
So far so well, however, there are few defects that need to be addressed in order to approve this project.

>  Some of your model's methods can be defined as association with SQL query instead of loading and processing objects in application memory _(as the database was created to store and query data it is always more efficient than ruby)_. Please avoid using enumerable methods where you should be using SQL query.

- [x] `User#pending_friends` can be defined as association with SQL query
```ruby
# User model

has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

- [x] `User#friend_requests` can be defined as association with SQL query
```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

- [x] - [ ] in order to get all posts written by a user and their friends, you can use a simple method in the User model (in order to NOT do it with Enumerable)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
- [x] Update your PR description after you have implemented these changes.

## Nice to have
Note that these are optional suggestions, which means you don't have to implement them. These are my recommendations.

- [ ] Keep your routes and views RESTful, e.g, instead of `#confirm_request` or `#reject_request` actions in your `FriendshipsController`, you could use `#update` and `#destroy` actions

Kindly implement these changes and resubmit for review.
Happy coding, and stay focused 🔋. I hope to see your next projects.

> #### <br>You can ping me on Slack @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) to draw my attention in case I do not respond quickly on here. I'm always available and will be glad to help.<br><br>

File: app/models/user.rb [hunk 1]
```diff
@@ -9,26 +9,33 @@ class User < ApplicationRecord
...
+  #  friends_array = []
 
-    inverse_friendships do |f1|
-      mirrored = false
-      friendships.map do |f2|
-        mirrored = true if f1.mirror?(f2)
-      end
+  #  friendships.map do |friendship|
+  #    friends_array << friendship.friend if friendship.confirmed
+  #  end
```
Comment (MahmoudBakr23) [inline app/models/user.rb]: Delete here!

File: app/models/user.rb [hunk 2]
```diff
@@ -9,26 +9,33 @@ class User < ApplicationRecord
...
 
-      friends_array << f1.friend unless mirrored || !f1.confirmed
-    end
+  #  inverse_friendships do |f1|
+  #    mirrored = false
+  #    friendships.map do |f2|
+  #      mirrored = true if f1.mirror?(f2)
+  #    end
```
Comment (MahmoudBakr23) [inline app/models/user.rb]: Delete here!

File: app/models/user.rb [hunk 3]
```diff
@@ -9,26 +9,33 @@ class User < ApplicationRecord
...
 
-    friends_array.compact
-  end
+  #    friends_array << f1.friend unless mirrored || !f1.confirmed
+  #  end
+
+  #  friends_array.compact
+  # end
```
Comment (MahmoudBakr23) [inline app/models/user.rb]: Delete here!

File: app/models/user.rb [hunk 4]
```diff
@@ -54,10 +61,17 @@ def friend?(user)
...
-    fposts.sort_by { |post| post[:created_at] }.reverse
+  # commented this old method for archive and learning purposes
+
+  # def friend_posts
+  #  fposts = []
+  #  friend_list.map { |friend| friend.posts.map { |post| fposts << post } }
+  #  posts.map { |post| fposts << post }
+  #  fposts.sort_by { |post| post[:created_at] }.reverse
+  # end
+
```
Comment (MahmoudBakr23) [inline app/models/user.rb]: Delete here!

File: app/models/friendship.rb [hunk 5]
```diff
@@ -6,8 +6,9 @@ class Friendship < ActiveRecord::Base
...
   # This filter needs to be deleted after mirror friend relationships have been handled in the database
-  def mirror?(friendship)
-    true if user.id == friendship.friend.id && friend.id == friendship.user.id
-    false
-  end
+  # commented this old method for archive and learning purposes
+  # def mirror?(friendship)
+  #  true if user.id == friendship.friend.id && friend.id == friendship.user.id
+  #  false
+  # end
```
Comment (MahmoudBakr23) [inline app/models/friendship.rb:13]: Delete here!

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Changes Required 🛑♻**

<hr>

Hello there :)
Great job working on this milestone and implementing the requested changes. However, there's one little thing left to do in order to get the approval 👍

<hr>

- [x] Please make sure that you delete any commented-out code from your models. See my comments below!

<hr>

Please apply the changes above, then ask for another CRR.
Good luck and happy coding 👨‍💻💻

**Comment by RokoVarano** [comment]
Comments have been deleted, however we are a bit concerned that this comments were actually for learning purposes, so we could see them later. The TSE's did not acknowledge this.

**Review by rukundoeric — APPROVED** [review]
# Status: 🍾 PROJECT APPROVED 🍾
_Excellent job on this project guys. 🎆 Keep it up._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve
  - n/a

_Congratulations on the splendid job **@team** 👍
You can now merge this PR and submit the project completion form 🚀 . I wish you all the very best in your next tasks 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/),
Slack: Rukundo Eric**

**Pull request merged**